src/content_main.js

```diff
@@ -100,46 +108,68 @@ async function trackVisitToSite() {
-			docHiddenTime = getCurrentTime();
-		} else {
-			// The page is now visible
+	document.addEventListener('visibilitychange', (event) => {
+		const tabIsVisible = !document.hidden;
+
+		if (tabIsVisible) {
 			docVisibleTime = getCurrentTime();
+		} else {
+			docHiddenTime = getCurrentTime();
 		}
 	});
 
 	/**
 	 * NOTE: setInterval may not be called while the document is hidden (while visibility lost)
 	 * https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Reasons_for_delays_longer_than_specified
 	 *
-	 * Store the recent time spent every 2 seconds to ensure time spent on site is recorded
-	 * even if the user closes the site.
+	 * Check if a new payment pointer is added to the page and store the recent time spent
+	 * every NEW_PAYMENT_POINTER_CHECK_RATE_MS milliseconds to ensure time spent on site is recorded
+	 * even if the user closes the site. Calls `handleValidationAndSetExtensionIcon` to set the extension icon.
 	 */
 	setInterval(async () => {
-		const now = getCurrentTime();
-
+		// Logic to store time spent at the origin (monetized or non-monetized)
 		if (document.visibilityState === 'visible') {
+			const currentTime = getCurrentTime();
+			let recentTimeSpent = 0;
+
 			if (docHiddenTime > previousStoreTime) {
 				// Adding time after visibility lost (document becomes hidden) and then gained (document is visible again)
 				// i.e. If the user navigates away from the site and then comes back
 				const timeFromPreviousStoreToDocHidden = docHiddenTime - previousStoreTime;
-				const timeSinceDocVisible = now - docVisibleTime;
-				await storeRecentTimeSpent(timeFromPreviousStoreToDocHidden + timeSinceDocVisible);
+				const timeSinceDocVisible = currentTime - docVisibleTime;
+				recentTimeSpent = timeFromPreviousStoreToDocHidden + timeSinceDocVisible;
 			} else {
 				// Adding time during regular interval (document visible)
-				await storeRecentTimeSpent(now - previousStoreTime);
+				recentTimeSpent = currentTime - previousStoreTime;
 			}
 
-			previousStoreTime = now;
+			// If the cached payment pointer is null, then the time elapsed is non-monetized time
+			// Otherwise, the recent time spent is monetized time
+			const isMonetizedTime = cachedPaymentPointer !== null;
+
+			await storeRecentTimeSpent(recentTimeSpent, isMonetizedTime);
+			previousStoreTime = currentTime;
+		}
+
+		// Logic to update the payment pointer and set the extension icon
+		const paymentPointerOnPage = getPaymentPointerFromPage();
+
+		if (paymentPointerOnPage !== cachedPaymentPointer) {
+			const isValidPaymentPointer = await validatePaymentPointer(paymentPointerOnPage);
+
+			// Set the extension icon to monetized if the payment pointer is valid
+			setExtensionIconMonetizationState(isValidPaymentPointer);
+
+			// Update the cached payment pointer if the payment pointer on the page has changed
+			cachedPaymentPointer = isValidPaymentPointer ? paymentPointerOnPage : null;
```

Comment:
@vezwork I went a different route than what you suggested in https://github.com/esse-dev/akita/pull/98#issuecomment-751951147, but it should accomplish the same thing.

Reply:
Don't we want the validation to occur though, since it could be validating a different payment pointer? I feel like we'd need to set up some sort of validation queue otherwise and probably integrate it with `isPaymentPointerRecentlyValidated()`?

Reply:
If we wait for validation of a payment pointer to finish, we still ensure that we validate all new/different payment pointers: If we wait for validation of a payment pointer to finish, it will then be set as the `cachedPaymentPointer` (if it is valid), and the next time the setinterval step happens, the new payment pointer on the page will not `=== cachedPaymentPointer` so it will then be validated. This method won't miss validating different payment pointers.

A validation queue sounds like it could work. Alternatively could have a variable `validatingPaymentPointer` which contains the payment pointer being validated, if a new/different payment pointer shows up, then you can start validating it otherwise you don't revalidate a payment pointer if it is the same as `validatingPaymentPointer`.

Reply:
I see, I interpreted your initial comment as "if validation is in progress already, check the flag and skip validation so we don't repeat it"

But you're actually saying, "if validation is in progress and it's for the same payment pointer, wait for the in progress validation to be done, instead of kicking off a new validation." Is my understanding correct?

Reply:
Yes!